library/core/src/mem/maybe_uninit.rs

```diff
@@ -319,9 +319,9 @@ impl<T> MaybeUninit<T> {
     /// Create a new array of `MaybeUninit<T>` items, in an uninitialized state.
     ///
     /// Note: in a future Rust version this method may become unnecessary
-    /// when array literal syntax allows
-    /// [repeating const expressions](https://github.com/rust-lang/rust/issues/49147).
-    /// The example below could then use `let mut buf = [MaybeUninit::<u8>::uninit(); 32];`.
+    /// when Rust allows
+    /// [inline const expressions](https://github.com/rust-lang/rust/issues/76001).
+    /// The example below could then use `let mut buf = [const { MaybeUninit::<u8>::uninit() }; 32];`.
```

Comment:
Isn't it already possible to write `let mut buf = [{ const U: MaybeUninit<u8> = MaybeUninit::uninit(); U}; 32];` thus making the method already unnecessary? Or is that too cumbersome to be considered usable?

Reply:
Yeah this is possible if you have the concrete type; it won't when generic types are involved.

Not sure what people prefer here, for now fixing the docs seemed easier than arguing for the method to be removed. ;)

Reply:
Ah yes, I missed the generic case.

Fixing the docs now is a good idea, I just encountered it earlier this week! But it was for a concrete type. ;)

Reply:
FYI @tspiteri that appears not to actually uninit.

```rust
#![feature(maybe_uninit_uninit_array)]

use std::mem::MaybeUninit;

pub fn f() -> [MaybeUninit<u8>; 32] {
    [{ const U: MaybeUninit<u8> = MaybeUninit::uninit(); U}; 32]
}

pub fn g() -> [MaybeUninit<u8>; 32] {
    MaybeUninit::uninit_array()
}
```

```asm
example::f:
        movq    %rdi, %rax
        xorps   %xmm0, %xmm0
        movups  %xmm0, 16(%rdi)
        movups  %xmm0, (%rdi)
        retq

example::g:
        movq    %rdi, %rax
        retq
```

Godbolt: https://rust.godbolt.org/z/8nnvPK1PM
A lot like https://github.com/rust-lang/rust/issues/83657

Reply:
> FYI @tspiteri that appears not to actually uninit.

Ah, good point, currently uninitialized `const`s are not actually uninitialized it seems...

Reply:
@RalfJung  What about @rodrimati1992's generic code here - for `[MaybeUninit<T>; N]`, which compiles on Rust 1.51 / part of arrayvec 0.7?  https://github.com/bluss/arrayvec/blob/7b290b7aa52b0c66b9056cfda577363d005f39b8/src/utils.rs

Reply:
Yep, that's where it was noticed 🙂